pulsar-broker/src/main/java/org/apache/pulsar/broker/web/RestEndpointMetricsFilter.java

```diff
@@ -0,0 +1,117 @@
+import com.google.common.cache.LoadingCache;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Histogram;
+import java.io.IOException;
+import java.util.Stack;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.container.ContainerResponseFilter;
+import javax.ws.rs.core.Response;
+import org.glassfish.jersey.server.internal.routing.UriRoutingContext;
+import org.glassfish.jersey.server.model.Resource;
+import org.glassfish.jersey.server.model.ResourceMethod;
+import org.jetbrains.annotations.NotNull;
+
+public class RestEndpointMetricsFilter implements ContainerResponseFilter, ContainerRequestFilter {
+    private static final LoadingCache<ResourceMethod, String> CACHE = CacheBuilder
+            .newBuilder()
+            .maximumSize(100)
+            .build(new CacheLoader<>() {
+                @Override
+                public @NotNull String load(@NotNull ResourceMethod method) throws Exception {
+                    return getRestPath(method);
+                }
+            });
+
+    private static final Histogram LATENCY = Histogram
+            .build("pulsar_broker_rest_endpoint_latency", "-")
+            .unit("ms")
+            .labelNames("path", "method")
+            .buckets(10D, 20D, 50D, 100D, 200D, 500D, 1000D, 2000D)
+            .register();
+    private static final Counter FAILED = Counter
+            .build("pulsar_broker_rest_endpoint_failed", "-")
+            .labelNames("path", "method", "code")
+            .register();
+
+    private static final String REQUEST_START_TIME = "requestStartTime";
+
+    @Override
+    public void filter(ContainerRequestContext req, ContainerResponseContext resp) throws IOException {
+        String path;
+        try {
+            UriRoutingContext info = (UriRoutingContext) req.getUriInfo();
+            ResourceMethod rm = info.getMatchedResourceMethod();
+            path = CACHE.get(rm);
+        } catch (Throwable ex) {
+            path = "UNKNOWN";
+        }
+
+        String method = req.getMethod();
+        Response.StatusType status = resp.getStatusInfo();
+        if (status.getStatusCode() < Response.Status.BAD_REQUEST.getStatusCode()) {
+            long start = req.getProperty(REQUEST_START_TIME) == null
+                    ? System.currentTimeMillis() : (long) req.getProperty(REQUEST_START_TIME);
+            LATENCY.labels(path, method).observe(System.currentTimeMillis() - start);
+        } else {
+            FAILED.labels(path, method, String.valueOf(status.getStatusCode())).inc();
+        }
+    }
+
+    @Override
+    public void filter(ContainerRequestContext req) throws IOException {
+        // Set the request start time into properties.
+        req.setProperty(REQUEST_START_TIME, System.currentTimeMillis());
+    }
+
+    private static String getRestPath(ResourceMethod method) {
```

Comment:
This is to converge the URI path, there are many rest-api usages in pulsar.
say, there is a endpoint which has a path `/api/v1/{id}/query`, the actual request path might be `/api/v1/111/query`, `/api/v1/222/query`,`/api/v1/333/query` and so on.
 it would lead to infinity label values and will broke the TSDB.

Reply:
I understand. This is why I wrote:

> If I understand correctly, this should give you to the matched path template of the Resource selected to execute against the HTTP request.

I wrote you can retrieve the path template defined by the JAX-Rs annotations using the code I wrote, no?

Reply:
I'd tested `containerRequestContext.getUriInfo().getMatchedURIs().get(0)`, it couldn't returns the URI template.
![image](https://github.com/apache/pulsar/assets/21362791/67074211-353b-4629-8acf-86f06492d931)